spec/core/ics-003-connection-semantics/UPGRADES.md

```diff
@@ -0,0 +1,365 @@
+
+The upgrade status path is a public path that can signal the status of the upgrade to the counterparty. It does not store anything in the successful case, but it will store a sentinel abort value in the case that a chain does not accept the proposed upgrade.
+
+```typescript
+function statusPath(id: Identifier): Path {
+    return "connections/{id}/upgradeStatus"
+
+}
+```
+
+### Upgrade Handshake
+
+The upgrade handshake defines four datagrams: *ConnUpgradeInit*, *ConnUpgradeTry*, *ConnUpgradeAck*, and *ConnUpgradeConfirm*
+
+A successful protocol execution flows as follows (note that all calls are made through modules per ICS 25):
+
+| Initiator | Datagram             | Chain acted upon | Prior state (A, B)          | Posterior state (A, B)      |
+| --------- | -------------------- | ---------------- | --------------------------- | --------------------------- |
+| Actor     | `ConnUpgradeInit`    | A                | (OPEN, OPEN)                | (UPGRADE_INIT, OPEN)        |
+| Actor     | `ConnUpgradeTry`     | B                | (UPGRADE_INIT, OPEN)        | (UPGRADE_INIT, UPGRADE_TRY) |
+| Relayer   | `ConnUpgradeAck`     | A                | (UPGRADE_INIT, UPGRADE_TRY) | (OPEN, UPGRADE_TRY)         |
+| Relayer   | `ConnUpgradeConfirm` | B                | (OPEN, UPGRADE_TRY)         | (OPEN, OPEN)                |
+
+At the end of an opening handshake between two chains implementing the sub-protocol, the following properties hold:
+
+- Each chain is running their new upgraded connection end and is processing upgraded logic and state according to the upgraded parameters.
+- Each chain has knowledge of and has agreed to the counterparty's upgraded connection parameters.
+
+If a chain does not agree to the proposed counterparty `UpgradeConnectionState`, it may abort the upgrade handshake by writing `UPGRADE_ERR` as the state in the `UpgradeStatus` and storing
+it under the status path and restore the original connection.
+
+`statusPath(id) => UpgradeStatus{UPGRADE_ERR}`
+
+A relayer may then submit a `CancelConnectionUpgradeMsg` to the counterparty. Upon receiving this message a chain must verify that the counterparty wrote `UPGRADE_ERR` into its `UpgradeStatus` and if successful, it will restore its original connection as well thus cancelling the upgrade.
+
+If an upgrade message arrives after the specified timeout, then the message MUST NOT execute successfully. Again a relayer may submit a proof of this in a `CancelConnectionUpgradeTimeoutMsg` so that counterparty cancels the upgrade and restores it original connection as well.
+
+```typescript
+function connUpgradeInit(
+    identifier: Identifier,
+    proposedUpgrade: UpgradeConnectionState
+) {
+    // current connection must be OPEN
+    currentConnection = provableStore.get(connectionPath(identifier))
+    abortTransactionUnless(connection.state == OPEN)
+
+    // abort transaction if an unmodifiable field is modified
+    // upgraded connection state must be in `UPGRADE_INIT`
```

Comment:
note that if we add fields to the connection struct this will by default allow them to be modified